BASE/src/Microsoft.ApplicationInsights/Extensibility/Implementation/Tracing/SelfDiagnosticsInternals/SelfDiagnosticsEventListener.cs

```diff
@@ -0,0 +1,125 @@
+        {
+            this.logLevel = logLevel;
+
+            // this.configRefresher = configRefresher ?? throw new ArgumentNullException(nameof(configRefresher));
+
+            List<EventSource> eventSources;
+            lock (this.lockObj)
+            {
+                eventSources = this.eventSourcesBeforeConstructor;
+                this.eventSourcesBeforeConstructor = null;
+            }
+
+            foreach (var eventSource in eventSources)
+            {
+#if NET452
+                this.EnableEvents(eventSource, this.logLevel, (EventKeywords)(-1));
+#else
+                this.EnableEvents(eventSource, this.logLevel, EventKeywords.All);
+#endif
+            }
+        }
+
+        internal void WriteEvent(string eventMessage, ReadOnlyCollection<object> payload)
+        {
+            // TODO
+        }
+
+        protected override void OnEventSourceCreated(EventSource eventSource)
+        {
+            if (ShouldSubscribe(eventSource))
+            {
+                // If there are EventSource classes already initialized as of now, this method would be called from
+                // the base class constructor before the first line of code in SelfDiagnosticsEventListener constructor.
+                // In this case logLevel is always its default value, "LogAlways".
+                // Thus we should save the event source and enable them later, when code runs in constructor.
+                if (this.eventSourcesBeforeConstructor != null)
+                {
+                    lock (this.lockObj)
+                    {
+                        if (this.eventSourcesBeforeConstructor != null)
+                        {
+                            this.eventSourcesBeforeConstructor.Add(eventSource);
+                            return;
+                        }
+                    }
+                }
+
+#if NET452
+                this.EnableEvents(eventSource, this.logLevel, (EventKeywords)(-1));
+#else
+                this.EnableEvents(eventSource, this.logLevel, EventKeywords.All);
+#endif
+            }
+
+            base.OnEventSourceCreated(eventSource);
+        }
+
+        /// <summary>
+        /// This method records the events from event sources to a local file, which is provided as a stream object by
+        /// SelfDiagnosticsConfigRefresher class. The file size is bound to a upper limit. Once the write position
+        /// reaches the end, it will be reset to the beginning of the file.
+        /// </summary>
+        /// <param name="eventData">Data of the EventSource event.</param>
+        protected override void OnEventWritten(EventWrittenEventArgs eventData)
+        {
+            this.WriteEvent(eventData.Message, eventData.Payload);
+        }
+
+        /// <summary>
+        /// This method checks if the given EventSource Name matches known EventSources that we want to subscribe to.
+        /// </summary>
+        private static bool ShouldSubscribe(EventSource eventSource)
```

Comment:
This method is copy-pasted from https://github.com/microsoft/ApplicationInsights-dotnet/blob/develop/BASE/src/Microsoft.ApplicationInsights/Extensibility/Implementation/Tracing/DiagnosticsModule/DiagnosticsEventListener.cs#L87